File: packages/vibechk/src/vibechk/runners.py
from __future__ import annotations

import subprocess
import sys
import time
from abc import ABC, abstractmethod
from dataclasses import dataclass
from pathlib import Path
from typing import TextIO

from vibechk.errors import VibechkError
from vibechk.models import Metrics, RunnerResult


@dataclass(frozen=True)
class RunnerRequest:
    plan_text: str
    workdir: Path
    model: str | None
    run_id: str
    artifacts_dir: Path


class Runner(ABC):
    @property
    @abstractmethod
    def name(self) -> str:
        """Stable runner name used by the CLI."""

    @abstractmethod
    def run(self, request: RunnerRequest, transcript_sink: TextIO) -> RunnerResult:
        """Run the agent and write raw output to transcript_sink."""


class RunnerRegistry:
    def __init__(self, runners: list[Runner]) -> None:
        self._runners = {runner.name: runner for runner in runners}

    def get(self, name: str) -> Runner:
        if name in self._runners:
            return self._runners[name]
        available = ", ".join(self.names()) or "none"
        raise VibechkError(f"Unsupported runner '{name}'. Available runners: {available}.")

    def names(self) -> list[str]:
        return sorted(self._runners)


class ClaudeRunner(Runner):
    @property
    def name(self) -> str:
        return "claude"

    def run(self, request: RunnerRequest, transcript_sink: TextIO) -> RunnerResult:
        command = self._command(request)
        started = time.monotonic()
        try:
            process = subprocess.Popen(
                command,
                cwd=request.workdir,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                text=True,
            )
        except FileNotFoundError as error:
            raise VibechkError("Runner 'claude' is not installed or not on PATH.") from error

        if process.stdout is not None:
            for line in process.stdout:
                sys.stdout.write(line)
                sys.stdout.flush()
                transcript_sink.write(line)
                transcript_sink.flush()

        exit_code = process.wait()
        wall_time_seconds = round(time.monotonic() - started, 3)
        return RunnerResult(
            exit_code=exit_code,
            transcript="",
            metrics=Metrics(wall_time_seconds=wall_time_seconds),
            artifacts={},
            runner_version=None,
        )

    def _command(self, request: RunnerRequest) -> list[str]:
        command = [
            "claude",
            "--bare",
            "--print",
            "--permission-mode",
            "acceptEdits",
        ]
        if request.model is not None:
            command.extend(["--model", request.model])
        command.append(request.plan_text)
        return command


def build_production_runner_registry() -> RunnerRegistry:
    return RunnerRegistry([ClaudeRunner()])

File: ts/packages/tools/vibechk/src/runners.ts
import { NodeCommandExecApi } from "@nseng-ai/foundation/exec";
import type { CommandExecApi, ExecResult } from "@nseng-ai/foundation/exec";

import { runVibechkCommand } from "./exec-util.ts";
import type { Metrics } from "./models.ts";
import { VibechkError } from "./store.ts";

export interface RunnerRequest {
	planText: string;
	workdir: string;
	model: string | null;
	runId: string;
	artifactsDir: string;
}

export interface RunnerResult {
	exitCode: number;
	transcript: string;
	metrics: Metrics;
	artifacts: Record<string, unknown>;
	runnerVersion: string | null;
}

export interface Runner {
	name: string;
	run(
		request: RunnerRequest,
		transcriptSink: (text: string) => void,
		stdoutSink: (text: string) => void,
	): Promise<RunnerResult>;
}

export class RunnerRegistry {
	private readonly runners: Map<string, Runner>;

	constructor(runners: readonly Runner[]) {
		this.runners = new Map(runners.map((runner) => [runner.name, runner]));
	}

	get(name: string): Runner {
		const runner = this.runners.get(name);
		if (runner !== undefined) {
			return runner;
		}
		const availableNames = Array.from(this.runners.keys()).sort().join(", ");
		const available = availableNames === "" ? "none" : availableNames;
		throw new VibechkError(`Unsupported runner '${name}'. Available runners: ${available}.`);
	}

	names(): readonly string[] {
		return Array.from(this.runners.keys()).sort();
	}
}

export class ClaudeRunner implements Runner {
	readonly name = "claude";
	private readonly execApi: CommandExecApi;

	constructor(execApi: CommandExecApi = new NodeCommandExecApi()) {
		this.execApi = execApi;
	}

	async run(
		request: RunnerRequest,
		transcriptSink: (text: string) => void,
		stdoutSink: (text: string) => void,
	): Promise<RunnerResult> {
		const command = this.buildCommand(request);
		const started = performance.now();

		const result = await this.executeCommand(command, request.workdir, transcriptSink, stdoutSink);
		const wallTimeSeconds = Math.round((performance.now() - started) / 10) / 100;

		return {
			exitCode: runnerExitCode(result),
			transcript: "",
			metrics: {
				wallTimeSeconds,
				inputTokens: null,
				outputTokens: null,
				totalTokens: null,
				costUsd: null,
			},
			artifacts: {},
			runnerVersion: null,
		};
	}

	private buildCommand(request: RunnerRequest): readonly [string, ...string[]] {
		const command = ["claude", "--print", "--permission-mode", "acceptEdits"];
		if (request.model !== null) {
			command.push("--model", request.model);
		}
		command.push(request.planText);
		return [command[0] ?? "claude", ...command.slice(1)];
	}

	private async executeCommand(
		command: readonly [string, ...string[]],
		workdir: string,
		transcriptSink: (text: string) => void,
		stdoutSink: (text: string) => void,
	): Promise<ExecResult> {
		const streamText = (text: string): void => {
			stdoutSink(text);
			transcriptSink(text);
		};

		return await runVibechkCommand({
			execApi: this.execApi,
			command: command[0],
			args: command.slice(1),
			execOptions: {
				cwd: workdir,
				onStdout: streamText,
				onStderr: streamText,
			},
			missingExecutableMessage: "Runner 'claude' is not installed or not on PATH.",
			startupFailurePrefix: "Runner 'claude' failed to start",
		});
	}
}

export function buildProductionRunnerRegistry(): RunnerRegistry {
	return new RunnerRegistry([new ClaudeRunner()]);
}

function runnerExitCode(result: ExecResult): number {
	switch (result.type) {
		case "spawn-failed":
			throw new VibechkError(`Runner 'claude' failed to start: ${result.error}`);
		case "cancelled":
			throw new VibechkError("Runner 'claude' was cancelled.");
		case "timed-out":
			throw new VibechkError("Runner 'claude' timed out.");
		case "exited":
			if (result.signal !== null) {
				throw new VibechkError(`Runner 'claude' was terminated by signal ${result.signal}.`);
			}
			if (result.code === null) {
				throw new VibechkError("Runner 'claude' exited without an exit code.");
			}
			return result.code;
	}
}
